lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstractions for system's DNF package information and repositories."""
+
+import copy
+import itertools
+import os
+import shutil
+import subprocess
+from collections import deque
+from concurrent.futures import ThreadPoolExecutor
+from enum import Enum
+from functools import wraps
+from io import StringIO
+from typing import Callable, Dict, List, Literal, Optional, Union
+
+
+class _PackageState(Enum):
+    INSTALLED = "installed"
+    AVAILABLE = "available"
+    ABSENT = "absent"
+
+
+class DNFExecutionError(Exception):
+    """Raise when dnf encounters an execution error."""
+
+
+def _dnf(
+    command: str,
+    args: Optional[Union[Union[str, os.PathLike], List[Union[str, os.PathLike]]]] = None,
+    preargs: Optional[List[str]] = None,
+    postargs: Optional[List[str]] = None,
+) -> subprocess.CompletedProcess[str]:
+    """Execute DNF command.
+
+    Args:
+        command (str): Command to execute.
+        args (Optional[Union[Union[str, os.PathLike], List[Union[str, os.PathLike]]]]):
+            Arguments to operate on. (Default: None).
+        preargs (Optional[List[str]], optional):
+            Optional arguments to prepend before command. (Default: None).
+        postargs (Optional[List[str]], optional):
+            Optional arguments to append after command. (Default: None).
+
+    Raises:
+        DNFExecutionError: Raised if specified DNF command fails to execute.
+
+    Returns:
+        (subprocess.CompletedProcess[str]):
+            returncode, stdout, and stderr of completed DNF command.
+    """
+    args = [args] if type(args) == str else args if type(args) == list else []
```

Comment:
I guess I should warn you that I commonly use one-liners because I like their brevity. You can find them all throughout my personal projects :). I understand your point though so I will make it more readable for the general population.

Reply:
I think you'll find they're just as unreadable for you in 12 months time. :-) I've often written dense code like that in the past, then come back to it later and been like "who on earth wrote this code?" then done a `git blame` ... oops, that was me!

Reply:
Removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb